packages/cli/cli/src/cli-context/upgrade-utils/getFernUpgradeMessage.ts

```diff
@@ -0,0 +1,162 @@
+import boxen from "boxen";
+import chalk from "chalk";
+import { FernUpgradeInfo } from "../CliContext";
+import { CliEnvironment } from "../CliEnvironment";
+import { FernGeneratorUpgradeInfo } from "./getGeneratorVersions";
+
+export function getFernUpgradeMessage({
+    cliEnvironment,
+    upgradeInfo
+}: {
+    cliEnvironment: CliEnvironment;
+    upgradeInfo: FernUpgradeInfo;
+}): string | undefined {
+    let upgradeAvailable = false;
+    let message = `${chalk.underline("Upgrades available")}\n\n`;
+
+    if (upgradeInfo.cliUpgradeInfo?.isUpgradeAvailable) {
+        message +=
+            "Fern " +
+            chalk.dim(cliEnvironment.packageVersion) +
+            chalk.reset(" → ") +
+            chalk.green(upgradeInfo.cliUpgradeInfo.latestVersion) +
+            " \n(Run " +
+            chalk.cyan(`${cliEnvironment.cliName} upgrade`) +
+            " to update)";
+        upgradeAvailable = true;
+    }
+
+    // To start we're truncating the list and recommending the user use a different command
+    // to see the full list so as to not overwhelm them
+    message += getGeneratorUpgradeMessage({
+        generatorUpgradeInfo: upgradeInfo.generatorUpgradeInfo,
+        limit: 2,
+        includeBoxen: false
+    });
+    const generatorsNeedingUpgrades = upgradeInfo.generatorUpgradeInfo.filter((gui) => gui.isUpgradeAvailable);
+    if (generatorsNeedingUpgrades.length > 0) {
+        message += `\nRun ${chalk.cyan("fern generator upgrade")} to upgrade your generators.`;
+        upgradeAvailable = true;
+    }
+    if (generatorsNeedingUpgrades.length > 2) {
+        message +=
+            `\nRun ${chalk.cyan("fern generator upgrade --list")}` +
+            " to see the full list of generator upgrades available.";
+    }
+
+    return upgradeAvailable
+        ? boxen(message, {
+              padding: 1,
+              float: "center",
+              textAlignment: "center",
+              borderColor: "yellow",
+              borderStyle: "round"
+          })
+        : undefined;
+}
+
+export function getGeneratorUpgradeMessage({
+    generatorUpgradeInfo,
+    header,
+    limit,
+    includeBoxen = true
+}: {
+    generatorUpgradeInfo: FernGeneratorUpgradeInfo[];
+    header?: string;
+    limit?: number;
+    includeBoxen?: boolean;
+}): string | undefined {
+    let message = header ?? "";
+    let generatorUpgradeNeeded = false;
+
+    let generatorsNeedingUpgrades = generatorUpgradeInfo.filter((gui) => gui.isUpgradeAvailable);
+    if (limit != null) {
+        generatorsNeedingUpgrades = generatorsNeedingUpgrades.slice(0, limit + 1);
+    }
+    for (const generatorUpgrade of generatorsNeedingUpgrades) {
```

Comment:
Based on the async implementation to get the latest generator versions, I think this list order will be non-deterministic. Not the end of the world since there's only a few entries, but it might be worth sorting the entries so that users always see the same result.

But if the test you added isn't flaky, then I could be wrong.